### PR TITLE
Add extension mechanism for dynamically discovering RAS error states

### PR DESCRIPTION
**Pull request by aravindksg**
Helps to address issue with adding any new RAS categories since existing mechanism is static array

### CODE REVIEW

**Comment by aravindksg** [comment]
@jandres742 @saik-intel @vaidyan1-int @wdamon-intel for review

draft for now as if this looks good, I can add back the new categories that were reverted. And also need to finish the other PR to have Sysman extension discovery mechanism

**Comment by aravindksg** [comment]
> @aravindksg Please create an issue to associate with this PR.

https://github.com/oneapi-src/level-zero-spec/issues/189

File: scripts/sysman/EXT_RasState.rst [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+========================
+ Ras Get State Extension
+========================
+
+API
+----
+
+* Functions
+
+    * ${s}RasGetStateExt
```
Comment (jandres742) [inline scripts/sysman/EXT_RasState.rst:24]: do we want this as an standard or experimental extension?

File: scripts/sysman/rasState.yml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      name: errorCounter
+      desc: "[out] Current value of RAS counter for specific error category."
+--- #--------------------------------------------------------------------------
+type: function
+desc: "Ras Get State"
+class: $sRas
+name: RasGetStateExt
+details:
+    - "This function retrieves error counters for different RAS error categories."
+    - "Clearing errors will affect other threads/applications - the counter values will start from zero."
```
Comment (jandres742) [inline scripts/sysman/rasState.yml:47]: who clears the errors?

Comment (jandres742) [reply scripts/sysman/rasState.yml:47]: ah, we have the clear below. So this function does two things: get errors, but also clear. Do we want to have two tasks in a single API, or better to have a different API to clear? Main issue with having both in the same function is that now it cannot be actually called from simultaneous threads, unless there's a lock inside the L0 driver.

by having two, we can indeed having multiple threads calling this getter, as all they would be doing is to retrieve informantion, w/o changing state.

Moreover, clear requires privileges?

File: scripts/sysman/EXT_RasState.rst (hunk 1 above)
Comment (aravindksg) [reply scripts/sysman/EXT_RasState.rst:24]: I prefer standard for this case as we want users to use this interface reliably going forward to gather Ras state struct for error categories. it's not intended to be experimental, but more definitive/stable mechanism to users

File: scripts/sysman/rasState.yml (hunk 2 above)
Comment (aravindksg) [reply scripts/sysman/rasState.yml:47]: sure, agree it's cleaner. the existing API for `zesRasGetState` does bundle it together and I just followed that here. but yes, I can separate this out.

Comment (aravindksg) [reply scripts/sysman/rasState.yml]: @jandres742 added separate ext to clear errors now.

File: scripts/sysman/rasState.yml [hunk 3]
```diff
@@ -0,0 +1,87 @@
+#
+# Copyright (C) 2023 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+#
+# See YaML.md for syntax definition
+#
+--- #--------------------------------------------------------------------------
+type: header
+desc: "Intel $OneApi Level-Zero Sysman Extension APIs for Ras Get State"
```
Comment (jandres742) [inline scripts/sysman/rasState.yml:10]: RAS

File: scripts/sysman/rasState.yml [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+version: "1.7"
+name: $S_RAS_GET_STATE_EXT_NAME
+value: '"$XS_extension_ras_state"'
+--- #--------------------------------------------------------------------------
+type: enum
+desc: "RAS Get State Extension Version(s)"
+version: "1.7"
+name: $s_ras_state_ext_version_t
+etors:
+    - name: "1_7"
```
Comment (jandres742) [inline scripts/sysman/rasState.yml:24]: this is the version of the extension, which is 1.0.

File: scripts/sysman/rasState.yml [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    - "This function clears error counters for a RAS error category."
+    - "Clearing errors will affect other threads/applications - the counter values will start from zero."
+    - "Clearing errors requires write permissions."
+    - "The application should not call this function from simultaneous threads."
+    - "The implementation of this function should be lock-free."
+params:
+    - type: $s_ras_handle_t
+      name: hRas
+      desc: "[in] Handle for the component."
+    - type: $x_bool_t
```
Comment (jandres742) [inline scripts/sysman/rasState.yml:79]: do we still need the boolean, given the function is called "Clear"?

File: scripts/sysman/rasState.yml (hunk 4 above)
Comment (aravindksg) [reply scripts/sysman/rasState.yml]: fixed

File: scripts/sysman/rasState.yml (hunk 3 above)
Comment (aravindksg) [reply scripts/sysman/rasState.yml]: fixed

File: scripts/sysman/rasState.yml (hunk 5 above)
Comment (aravindksg) [reply scripts/sysman/rasState.yml]: fixed

File: scripts/sysman/rasState.yml [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      name: pCount
+      desc: |
+            [in,out] pointer to the number of RAS state structures that can be retrieved.
+            if count is zero, then the driver shall update the value with the total number of error categories for which state can be retrieved.
+            if count is greater than the number of RAS states available, then the driver shall update the value with the correct number of RAS states available.
+    - type: $s_ras_state_t*
+      name: pState
+      desc: |
+            [in,out][optional][range(0, *pCount)] array of query results for RAS error states for different categories.
+            if count is less than the number of RAS states available, then driver shall only retrieve that number of RAS states.
```
[the remark this answers is not in the document]
Comment (wdamon-intel) [reply scripts/sysman/rasState.yml:63]: @mraghuwa The `zes_ras_state_t::category` type cannot change as that has the potential to break backwards compatibility (even if it is the same size, the meaning, interpretation, magnitude and wrap-around behavior changes). Perhaps this extension could include a query for the supported categories, but really the legacy path should be deprecated, and the user should be encouraged to move to this extension.

File: scripts/sysman/rasState.yml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    - type: $s_ras_handle_t
+      name: hRas
+      desc: "[in] Handle for the component."
+    - type: uint32_t*
+      name: pCount
+      desc: |
+            [in,out] pointer to the number of RAS state structures that can be retrieved.
+            if count is zero, then the driver shall update the value with the total number of error categories for which state can be retrieved.
+            if count is greater than the number of RAS states available, then the driver shall update the value with the correct number of RAS states available.
+    - type: $s_ras_state_t*
```
[the remark this answers is not in the document]
Comment (aravindksg) [reply scripts/sysman/rasState.yml]: done

File: scripts/sysman/EXT_RasState.rst [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+* Structures
+
+    * ${s}_ras_state_ext_t
+
+
+Ras State
```
Comment (aravindksg) [inline scripts/sysman/EXT_RasState.rst:29]: @wdamon-intel  @jandres742  added this section to document intentions for extension and also some pseudo code. please check

File: scripts/sysman/rasState.yml [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+--- #--------------------------------------------------------------------------
+type: header
+desc: "Intel $OneApi Level-Zero Sysman Extension APIs for RAS Get State and Clear State"
+version: "1.7"
+--- #--------------------------------------------------------------------------
+type: macro
+desc: "RAS Get State Extension Name"
+version: "1.7"
+name: $S_RAS_GET_STATE_EXT_NAME
+value: '"$XS_extension_ras_state"'
```
Comment (jandres742) [inline scripts/sysman/rasState.yml:17]: I dont think XS is correct. right @wdamon-intel ?

Comment (aravindksg) [reply scripts/sysman/rasState.yml:17]: followed what we have in powerLimits.yml . let me know if it needs to be fixed

Comment (wdamon-intel) [reply scripts/sysman/rasState.yml:17]: @jandres742 @aravindksg XS is correct.

File: scripts/sysman/rasState.yml [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      name: "category"
+      desc: "[out] category for which error counter is provided."
+    - type: uint64_t
+      name: errorCounter
+      desc: "[out] Current value of RAS counter for specific error category."
+--- #--------------------------------------------------------------------------
+type: function
+desc: "Ras Get State"
+class: $sRas
+name: RasGetStateExp
```
Comment (jandres742) [inline scripts/sysman/rasState.yml:44]: why double 
Ras? in the class and in the name?

Comment (aravindksg) [reply scripts/sysman/rasState.yml]: oops, fixed both here and for clear state

**Review by jandres742 — APPROVED** [review]
@wdamon-intel please check

**Comment by aravindksg** [comment]
> @wdamon-intel please check

@jandres742  @wdamon-intel @sanilkumar0 @vaidyan1-int added new enum for RAS category to this PR since we would be using this new category via new extension `RasGetStateExp`. please re-review

**Review by jandres742 — APPROVED** [review]

**Comment by jandres742** [comment]
So are we deprecating the other interface and enums? if so, where are we adding that?

**Review by wdamon-intel — APPROVED** [review]

**Comment by wdamon-intel** [comment]
> So are we deprecating the other interface and enums? if so, where are we adding that?

Suggest deprecating the old interface in a future point release (v1.8 or v1.9) so that this experimental extension has the necessary time to mature.

File: scripts/sysman/EXT_Exp_RasState.rst [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+In L0 Sysman, we report these errors via RAS API and organized into Correctable or Uncorrectable
+errors. Furthermore, the errors can be grouped by category so that source of the error
+is easily identifiable.
+
+This extension is defined with intent to provide an extensible interface to the user for discovering
+these errors. A separate function is provided to allow users to clear the error counters.
+This functionality extends and is intended to eventually replace existing mechanism via ${s}RasGetState.
+
+To that end, this extension also adds new enumerator for RAS error categories. The list of error categories
+include existing ones (refer  Sysman Programming guide) and also additional ones which are defined here.
```
Comment (wdamon-intel) [inline scripts/sysman/EXT_Exp_RasState.rst:47]: "refer to the Sysman Programming Guide"

File: scripts/sysman/EXT_Exp_RasState.rst [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+The additional error categories are listed below: 
+
++-------------------------------------------------------------------+--------------------------------------------------------------+----------------------------------------------------------------+
+| Error category                                                    | ${S}_RAS_ERROR_TYPE_CORRECTABLE | ${S}_RAS_ERROR_TYPE_UNCORRECTABLE |
++===================================================================+==============================================================+================================================================+
+| ${S}_RAS_ERROR_CAT_MEMORY_ERRORS                                   | Number of ECC                                                | Number of ECC                                                  |
```
Comment (wdamon-intel) [inline scripts/sysman/EXT_Exp_RasState.rst:53]: This programming guide should use the new enum type values, `${S}_RAS_ERROR_CATEGORY_EXP_MEMORY_ERRORS`

File: scripts/sysman/EXT_Exp_RasState.rst [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+|                                                                   | correctable errors                                           | uncorrectable errors                                           |
+|                                                                   | that have occurred in                                        | that have occurred in                                          |
+|                                                                   | memory - GDDR/HBM.                                           | memory - GDDR/HBM                                              |
++-------------------------------------------------------------------+--------------------------------------------------------------+----------------------------------------------------------------+
+| ${S}_RAS_ERROR_CAT_SCALE_ERRORS                                    | Number of                                                    | Number of ECC                                                  |
```
Comment (wdamon-intel) [inline scripts/sysman/EXT_Exp_RasState.rst:58]: same

File: scripts/sysman/EXT_Exp_RasState.rst [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+|                                                                   | correctable errors                                           | uncorrectable errors                                           |
+|                                                                   | that have occurred in                                        | that have occurred in                                          |
+|                                                                   | scale fabric.                                                | scale fabric.                                                  |
++-------------------------------------------------------------------+--------------------------------------------------------------+----------------------------------------------------------------+
+| ${S}_RAS_ERROR_CAT_L3FABRIC_ERRORS                                 | Number of ECC                                                | Number of ECC                                                  |
```
Comment (wdamon-intel) [inline scripts/sysman/EXT_Exp_RasState.rst:63]: same

**Review by wdamon-intel — CHANGES_REQUESTED** [review]

File: scripts/sysman/EXT_Exp_RasState.rst (hunk 11 above)
Comment (aravindksg) [reply scripts/sysman/EXT_Exp_RasState.rst]: fixed

File: scripts/sysman/EXT_Exp_RasState.rst (hunk 12 above)
Comment (aravindksg) [reply scripts/sysman/EXT_Exp_RasState.rst]: fixed

File: scripts/sysman/EXT_Exp_RasState.rst (hunk 13 above)
Comment (aravindksg) [reply scripts/sysman/EXT_Exp_RasState.rst]: fixed

File: scripts/sysman/EXT_Exp_RasState.rst (hunk 14 above)
Comment (aravindksg) [reply scripts/sysman/EXT_Exp_RasState.rst]: fixed

**Review by wdamon-intel — APPROVED** [review]

**Review by jandres742 — APPROVED** [review]

**Pull request merged**